[query] Truncate search results to prevent OOM on jaeger-query.

annanay25 opened the pull request:
Resolves #1051.

Changes:
- Added upper bound on number of traces to fetch from a Cassandra backend.
- In addition to upper bound on traces, added upper bound on number of spans to fetch from an ES backend.

Question: Do we need a flag to mark the result as truncated on the UI? 

Signed-off-by: Annanay <annanay.a@media.net>

yurishkuro commented on plugin/storage/cassandra/spanstore/reader.go:
```diff
@@ -237,7 +237,10 @@ func (s *SpanReader) FindTraces(ctx context.Context, traceQuery *spanstore.Trace
 	if traceQuery.NumTraces == 0 {
 		traceQuery.NumTraces = defaultNumTraces
+	} else if traceQuery.NumTraces > 10000 {
+		traceQuery.NumTraces = 10000
```
why this hardcoded upper limit? It should be user-configurable.

annanay25 replied:
^ This is fine. My bad, confused the two.

annanay25 commented:
So I'm a little confused. Each query can already be parameterised with a limit on number of traces - https://github.com/jaegertracing/jaeger/blob/master/cmd/query/app/query_parser.go#L139 (if not specified then a default is used) - doesn't this resolve jaegertracing/jaeger#1051? If no, then why not?

yurishkuro commented:
The original ticket was about individual traces with large number of spans (we've see traces with tens of millions of spans).

annanay25 commented:
@yurishkuro OK.

In that case, I will drop the following changes to `plugin/storage/cassandra/spanstore/reader.go`

```
	if traceQuery.NumTraces == 0 {
		traceQuery.NumTraces = defaultNumTraces
	} else if traceQuery.NumTraces > 10000 {
		traceQuery.NumTraces = 10000
        }
```

But will retain the following `TerminateAfter(defaultNumSpans)` - 

```
searchRequests[i] = elastic.NewSearchRequest().IgnoreUnavailable(true).Type(spanType).Source(elastic.NewSearchSource().Query(query).Size(defaultDocCount).TerminateAfter(defaultNumSpans).Sort("startTime", true).SearchAfter(nextTime))
```

Should `defaultNumSpans` be user configurable? This would a change in the parameter list of the function `multiRead()`.

annanay25 commented:
PR updated. 

Need to figure out a way parameterise `GetTrace()` (would require changes to the `spanstore.Reader` interface!) with `numSpans` if a user wants to force retrieval of all spans.

annanay25 commented:
@yurishkuro could you please re-run Travis on this?

pavolloffay commented:
There is a git conflict.It needs to be rebased.

annanay25 commented:
@pavolloffay I've rebased and removed conflicts. 

Also, if we want to allow users to force retrieval of more than 10k spans for a trace, we'd have to make changes to the `spanstore.Reader` interface. What's our take on that?

annanay25 commented on storage/spanstore/interface.go:
```diff
@@ -51,4 +51,5 @@ type TraceQueryParameters struct {
 	DurationMin   time.Duration
 	DurationMax   time.Duration
 	NumTraces     int
+	NumSpans      int
```
Hi @yurishkuro, I think this particular issue was filed for OOM on the jaeger-query nodes, and we have a separate issue for handling overloading of the underlying storage - #960

annanay25 replied:
OK. Will look into. Closing this PR.

So do we want to restrict number of spans (per trace) at ingestion?

yurishkuro replied:
Not at ingestion (which isn't possible anyway since spans arrive async). We want the limit to be applied at query time, similar to the changes you made to `plugin/storage/es/spanstore/reader.go`, but the value of the limit to be configurable as a parameter of the query-service, passed at Reader construction from CLI, not as a parameter of the HTTP request.

annanay25 replied:
Sorry its taking so many iterations to get this one right. Understood now, sending in a new, clean PR.
